src/afrolabs/components/kafka.clj

```diff
@@ -1547,6 +1559,74 @@
                            (s/keys :req-un [::ktable-id]
                                    :opt-un [::caught-up-once?])))
 
+(defprotocol IKTable
+  (ktable-wait-for-catchup
+    [_this topic-partition-offset timeout-period timeout-value]
+    "Waits until a ktable has consumed messages from its source topics, at least up to the topic-partition-offset data parameter.
+Supports a timeout operation."))
+
+(defn ktable-atom-wait-for-catchup
+  "Will use csp to actually wait up to timeout-duration for the ktable value to reflect changes up to this level."
+  [ktable-atom topic-partition-offset timeout-duration timeout-value]
+  (let [timeout-ms (.toMillis ^java.time.Duration timeout-duration)
+        timeout-chan (csp/timeout timeout-ms)
+
+        initial-ktable-value @ktable-atom
+
+        ktable->topic-partition-offsets
+        (fn [ktable-value] (:ktable/topic-partition-offsets (meta ktable-value)))
+
+        caught-up-already? (fn [ktable-value]
+                             (let [ktable-topic-partition-offsets
+                                   (ktable->topic-partition-offsets ktable-value)]
+
+                               (->> topic-partition-offset
+                                    (mapcat (fn [[topic partition-offset]]
+                                              (map (fn [[partition offset]] [topic partition offset])
+                                                   partition-offset)))
+                                    (keep (fn [[topic partition offset]]
+                                            (let [ktable-progress-offset (get-in ktable-topic-partition-offsets
+                                                                                 [topic partition])]
+                                              (when (< ktable-progress-offset offset)
+                                                :not-caught-up-yet))))
+                                    (count)
+                                    (zero?))))]
+    (cond
+      (caught-up-already? initial-ktable-value)
+      (ktable->topic-partition-offsets initial-ktable-value)
```

Comment:
Early return if we start out knowing about those topics-partition-offsets already.